# module4_RNAseq
The fourth module of the 2018 Bioinformatics Community of Practice at BecA-ILRI Hub


<img src="figures/Overview.svg?raw=true" width="100%" >

## Overview of RNA-seq
We will present an overview of how the RNA-Seq samples are produced and the drawbacks of aligning the reads naively to a genomic reference

In the exercises we will explore the test dataset. 

## Mapping reads to a known transcriptome
We will use a mini test set to align the reads to a) the transcriptome (kallisto) and b) The reference with the annotation. 

## SNP calling and VCF files
We will use the BAM files from b) to call for SNPs with freebayes  (or samtools directly?)


## Differential expression
We will explore different R packages for Differential expression (please state the ones you are planning to use). 

## Transcriptome assembly
We will use the same reads to make a de-novo assembly and produce a putative functional annotation (trinity and cufflinks)

## Gene Ontology. 
We will use TRAPID to annotate some genes.
